orttraining/orttraining/training_ops/cuda/communication/nccl_service.cc

```diff
@@ -0,0 +1,346 @@
+};
+
+void NcclService::PlanNewGroupStart() {
+  group_status_.push_back(true);
+  schedule_.push_back(NcclTaskGroup());
+};
+
+void NcclService::PlanNewGroupEnd() {
+  group_status_.back() = false;
+};
+
+void NcclService::PlanSend(const int dst) {
+  ORT_ENFORCE(group_status_.back(), "Last communication group can not be changed after call PlanEndNewGroup.");
+
+  schedule_.back().PlanTask(NcclTask::Type::SEND, {dst});
+};
+
+void NcclService::PlanRecv(const int src) {
+  ORT_ENFORCE(group_status_.back(), "Last communication group can not be changed after call PlanEndNewGroup.");
+  schedule_.back().PlanTask(NcclTask::Type::RECV, {src});
+};
+
+void NcclService::WaitForLaunch() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  cv_.wait(lock, [this] { return is_running_; });
+}
+
+std::ostream& operator<<(std::ostream& stream, const NcclTaskGroup& task_group) {
+  for (int i = 0; static_cast<size_t>(i) < task_group.batch.size(); ++i) {
+    std::string line = "  ";
+    auto& task = task_group.batch[i];
+    if (task.type == NcclTask::Type::SEND) {
+      line += "Send, ";
+    } else if (task.type == NcclTask::Type::RECV) {
+      line += "Recv, ";
+    }
+
+    for (int j = 0; static_cast<size_t>(j) < task.peers.size(); ++j) {
+      line += std::to_string(task.peers[j]);
+      if (static_cast<size_t>(j) != task.peers.size() - 1) {
+        line += ", ";
+      } else {
+        line += "\n";
+      }
+    }
+    stream << line;
+  }
+  return stream;
+}
+
+std::ostream& operator<<(std::ostream& stream, const NcclService& service) {
+  for (int i = 0; static_cast<size_t>(i) < service.schedule_.size(); ++i) {
+    stream << "NCCL operations at time " << i << std::endl;
+    stream << service.schedule_[i];
+  }
+  return stream;
+}
+
+int NcclService::FindNextCommunicationTime() const {
+  for (int i = 0; static_cast<size_t>(i) < schedule_.size(); ++i) {
+    if (schedule_[i].IsAllTasksEqueued() && !schedule_[i].IsAllTasksFinished()) {
+      return i;
+    }
+  }
+  return -1;
+};
+
+void NcclService::SubmitSendAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  WaitForLaunch();
+  auto& profile_context = profile::Context::GetInstance();
+  const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+
+  // Pointer to enqueued task.
+  const NcclTask* task;
+
+  // Submit task.
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& profile_context = profile::Context::GetInstance();
+    const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+    task = schedule_[time_].EqueueTask(NcclTask::Type::SEND, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&] { return task->is_finished; });
+  }
+};
+
+void NcclService::SubmitRecvAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  WaitForLaunch();
```

Comment:
>WaitForLaunch(); [](start = 2, length = 16)

Do we need this `WaitForLaunch()` for every `SubmitSendAndWait()` and `SubmitRecvAndWait()`? `is_running_` is set in main thread `launch()` before worker thread starts. So by the time ORT starts training, `is_running_` should already set to True? 

As `WaitForLaunch()` constantly acquire the lock, and lock acquiring usually is very expensive, I am thinking it'll be better if we can avoid it as much as possible.